File: HungryTools/Classes/ZHCTextField/ZHCTextField.m
//
//  ZHCTextField.m
//  HungryTools
//
//  Created by 张海川 on 2019/4/25.
//

#ifndef ThemeColor
#define ThemeColor [UIColor colorWithRed:255/255.f green:80/255.f blue:74/255.f alpha:1]
#endif

#import "ZHCTextField.h"
#import "Masonry.h"

#pragma mark - NSString Category

@interface NSString (HGZCheck)

@end

@implementation NSString (HGZCheck)

- (BOOL)hgz_checkWithRegexString:(NSString *)regexString {
    return [[NSPredicate predicateWithFormat:@"SELF MATCHES %@", regexString] evaluateWithObject:self];
}

@end

#pragma mark - ZHCTextField

@interface ZHCTextField() <UITextFieldDelegate>

@property (nonatomic, strong) UIView *      bottomLineView;
@property (nonatomic, strong) UIView *      secureView;
@property (nonatomic, strong) UIButton *    secureButton;

@property (nonatomic, assign) BOOL          isContainLeftOrRightView;

@end

@implementation ZHCTextField

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        [self initConfig];
    }
    return self;
}

- (void)initConfig {
    self.clearButtonMode = UITextFieldViewModeWhileEditing;
    self.delegate = self;
    self.font = [UIFont systemFontOfSize:14];
    
    _isContainLeftOrRightView = NO;
    _bottomLineHeight = 0.5;
    
    [self addBottomLine];
}

// 无效
//- (CGRect)clearButtonRectForBounds:(CGRect)bounds {
//    //        bounds.origin.x -= ViewWidth(self.rightView);
//    CGRect frame = [super clearButtonRectForBounds:bounds];
//    if (self.rightView) {
//        frame.origin.x -= ViewWidth(self.rightView) + 10;
//    }
//    return frame;
//}

#pragma mark - UITextFieldDelegate

- (BOOL)textField:(UITextField *)textField shouldChangeCharactersInRange:(NSRange)range replacementString:(NSString *)string {
    // 删除
    if ([string isEqualToString:@""]) {
        return YES;
    }
    
    NSInteger comingTextLength = textField.text.length + string.length - range.length;
    
    if (_maxLength) {
        if (comingTextLength > _maxLength) {
            return NO;
        }
    }
    
    switch (_fieldType) {
        case ZHCFieldTypeDefault: {
            return YES;
            break;
        }
        case ZHCFieldTypeNumber: {
            return [string hgz_checkWithRegexString:@"[0-9]+"];
            break;
        }
        case ZHCFieldTypePhoneNumber: {
            if (self.text.length == 3 || self.text.length == 8) {
                self.text = [self.text stringByAppendingString:@" "];
            }
            return [string hgz_checkWithRegexString:@"[0-9]+"];
            break;
        }
        case ZHCFieldTypePhoneNumberWithoutSpacing: {
            return [string hgz_checkWithRegexString:@"[0-9]+"];
            break;
        }
        case ZHCFieldTypePassword: {
            // 半角字符 包括字母，数字，标点符号
            return [string hgz_checkWithRegexString:@"[\\x00-\\xff]+"];
            break;
        }
        case ZHCFieldTypeMoney: {
            return [string hgz_checkWithRegexString:@"[0-9.]+"];
            break;
        }
        case ZHCFieldTypeIDCardNumber: {
            return [string hgz_checkWithRegexString:@"[0-9Xx]+"];
            break;
        }
        case ZHCFieldTypeChinese: {
            return [string hgz_checkWithRegexString:@"[a-z\\u4e00-\\u9fa5]+"];
            break;
        }
        case ZHCFieldTypeBankCardNumber: {
            return [string hgz_checkWithRegexString:@"[0-9]+"];
            break;
        }
    }
    return NO;
}

- (void)textFieldDidBeginEditing:(UITextField *)textField {
    [UIView animateWithDuration:0.25 animations:^{
        if (self.bottomLineActiveColor) {
            self.bottomLineView.backgroundColor = self.bottomLineActiveColor;
        } else {
            self.bottomLineView.backgroundColor = ThemeColor;
        }
    }];
}

- (void)textFieldDidEndEditing:(UITextField *)textField {
    [UIView animateWithDuration:0.25 animations:^{
        if (self.bottomLinePassiveColor) {
            self.bottomLineView.backgroundColor = self.bottomLinePassiveColor;
        } else {
            self.bottomLineView.backgroundColor = [UIColor lightGrayColor];
        }
    }];
}

#pragma mark - Action

- (void)secureBtnClicked {
    self.secureTextEntry ^= 1;
    self.secureButton.selected = !self.isSecureTextEntry;
}

#pragma mark - Private Method

- (void)addBottomLine {
    
    [self addSubview:self.bottomLineView];
    [self.bottomLineView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.bottom.right.mas_equalTo(0);
        make.height.mas_equalTo(self.bottomLineHeight);
    }];
}

- (void)updateBottomLineConstraints {
    
    _isContainLeftOrRightView = YES;
    
    [self.bottomLineView mas_remakeConstraints:^(MASConstraintMaker *make) {
        make.left.bottom.right.mas_equalTo(UIEdgeInsetsMake(0, 10, 0, 10));
        make.height.mas_equalTo(self.bottomLineHeight);
    }];
}

- (void)setPlaceholderAttribute:(id)value {
    
    NSMutableAttributedString * attriString = [[NSMutableAttributedString alloc] initWithAttributedString:self.attributedPlaceholder];
    
    NSAttributedStringKey name;
    if ([value isKindOfClass:UIFont.class]) {
        name = NSFontAttributeName;
    } else if ([value isKindOfClass:UIColor.class]) {
        name = NSForegroundColorAttributeName;
    }
    [attriString addAttribute:name value:value range:NSMakeRange(0, self.placeholder.length)];
    
    self.attributedPlaceholder = attriString;
}

#pragma mark - Public Method

#pragma mark - Setter

- (void)setFieldType:(ZHCFieldType)fieldType {
    
    _fieldType = fieldType;
    
    switch (_fieldType) {
        case ZHCFieldTypeDefault: {
            
            break;
        }
        case ZHCFieldTypeNumber: {
            self.keyboardType = UIKeyboardTypeNumberPad;
            break;
        }
        case ZHCFieldTypePhoneNumber: {
            self.keyboardType = UIKeyboardTypePhonePad;
            self.maxLength = 13;
            break;
        }
        case ZHCFieldTypePhoneNumberWithoutSpacing: {
            self.keyboardType = UIKeyboardTypePhonePad;
            self.maxLength = 11;
            break;
        }
        case ZHCFieldTypePassword:{
            self.secureTextEntry = YES;
            self.rightView = self.secureView;
            self.rightViewMode = UITextFieldViewModeAlways;
            self.keyboardType = UIKeyboardTypeAlphabet;
            self.maxLength = 18;
            break;
        }
        case ZHCFieldTypeMoney: {
            self.keyboardType = UIKeyboardTypeDecimalPad;
            self.maxLength = 8;
            break;
        }
        case ZHCFieldTypeIDCardNumber: {
            self.maxLength = 18;
            break;
        }
        case ZHCFieldTypeChinese: {
            
            break;
        }
        case ZHCFieldTypeBankCardNumber: {
            self.keyboardType = UIKeyboardTypeNumberPad;
            self.maxLength = 19;
            break;
        }
    }
}

- (void)setLeftView:(UIView *)leftView {
    [super setLeftView:leftView];
    [self updateBottomLineConstraints];
}

- (void)setRightView:(UIView *)rightView {
    [super setRightView:rightView];
    [self updateBottomLineConstraints];
}

- (void)setLeftText:(NSString *)leftText {
    
    UILabel * textLabel = [UILabel new];
    textLabel.font = [UIFont systemFontOfSize:14];
    textLabel.text = leftText;
    if (_leftTextColor) {
        textLabel.textColor = _leftTextColor;
    }
    if (_leftTextFontSize > 0) {
        textLabel.font = [UIFont systemFontOfSize:_leftTextFontSize];
    }
    
    CGFloat textWidth = [textLabel.text sizeWithAttributes:@{NSFontAttributeName : textLabel.font}].width;
    UIView * leftView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, textWidth + 21, self.bounds.size.height)];
    
    [leftView addSubview:textLabel];
    [textLabel mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerY.mas_equalTo(0);
        make.left.mas_equalTo(10);
        make.right.mas_equalTo(-10);
    }];
    
    self.leftView = leftView;
    self.leftViewMode = UITextFieldViewModeAlways;
}

- (void)setLeftImageName:(NSString *)leftImageString {
    
    UIImage * image = [UIImage imageNamed:leftImageString];
    UIImageView * subView = [[UIImageView alloc] initWithImage:image];
    
    UIView * leftView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, image.size.width + 20, self.bounds.size.height)];
    
    [leftView addSubview:subView];
    [subView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerY.mas_equalTo(0);
        make.left.mas_equalTo(10);
        make.right.mas_equalTo(-10);
    }];
    
    self.leftView = leftView;
    self.leftViewMode = UITextFieldViewModeAlways;
}

- (void)setLeftImage:(UIImage *)leftImage {
    
    UIImageView * subView = [[UIImageView alloc] initWithImage:leftImage];
    
    UIView * leftView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, leftImage.size.width + 20, self.bounds.size.height)];
    
    [leftView addSubview:subView];
    [subView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerY.mas_equalTo(0);
        make.left.mas_equalTo(10);
        make.right.mas_equalTo(-10);
    }];
    
    self.leftView = leftView;
    self.leftViewMode = UITextFieldViewModeAlways;
}

- (void)setShowBottomLine:(BOOL)isShowBottomLine {
    
    _showBottomLine = isShowBottomLine;
    self.bottomLineView.hidden = !_showBottomLine;
}

- (void)setSecureButtonImages:(NSArray<UIImage *> *)secureButtonImages {
    if (secureButtonImages.count != 2) {
        return;
    }
    if ([secureButtonImages[0] isKindOfClass:[UIImage class]]) {
        [_secureButton setImage:secureButtonImages[0] forState:UIControlStateSelected];
    }
    if ([secureButtonImages[1] isKindOfClass:[UIImage class]]) {
        [_secureButton setImage:secureButtonImages[1] forState:UIControlStateNormal];
    }
}

- (void)setBottomLinePassiveColor:(UIColor *)bottomLinePassiveColor {
    _bottomLinePassiveColor = bottomLinePassiveColor;
    self.bottomLineView.backgroundColor = _bottomLinePassiveColor;
}

- (void)setBottomLineHeight:(float)bottomLineHeight {
    
    _bottomLineHeight = bottomLineHeight;
    
    [self.bottomLineView mas_remakeConstraints:^(MASConstraintMaker *make) {
        if (self.isContainLeftOrRightView) {
            make.left.bottom.right.mas_equalTo(UIEdgeInsetsMake(0, 10, 0, 10));
        } else {
            make.left.bottom.right.mas_equalTo(0);
        }
        make.height.mas_equalTo(bottomLineHeight);
    }];
}

- (void)setClearButtonImage:(UIImage *)clearButtonImage {
    _clearButtonImage = clearButtonImage;
    
    UIButton *button =  [self valueForKey:@"_clearButton"];
    [button setImage:clearButtonImage forState:UIControlStateNormal];
}

- (void)setPlaceholderColor:(UIColor *)placeholderColor {
    _placeholderColor = placeholderColor;
    [self setPlaceholderAttribute:placeholderColor];
}

- (void)setPlaceholderFont:(UIFont *)placeholderFont {
    _placeholderFont = placeholderFont;
    [self setPlaceholderAttribute:placeholderFont];
}

- (void)setPlaceholder:(NSString *)placeholder {
    [super setPlaceholder:placeholder];
    
    if (_placeholderColor) {
        [self setPlaceholderAttribute:_placeholderColor];
    }
    if (_placeholderFont) {
        [self setPlaceholderAttribute:_placeholderFont];
    }
}

#pragma mark - Getter

- (NSString *)phoneNumberString {
    
    NSString * string;
    
    if (self.fieldType == ZHCFieldTypePhoneNumber) {
        string = [self.text stringByReplacingOccurrencesOfString:@" " withString:@""];
    }
    
    return string;
}

- (UIView *)bottomLineView {
    if (!_bottomLineView) {
        _bottomLineView = [UIView new];
        _bottomLineView.backgroundColor = [UIColor colorWithRed:216/255.f green:216/255.f blue:216/255.f alpha:1];
    }
    return _bottomLineView;
}

- (UIView *)secureView {
    if (!_secureView) {
        _secureView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, 64, 30)];
        
        self.secureButton.frame = CGRectMake(10, 0, 44, 30);
        [_secureView addSubview:self.secureButton];
    }
    return _secureView;
}

- (UIButton *)secureButton {
    if (!_secureButton) {
        _secureButton = [UIButton buttonWithType:UIButtonTypeCustom];
        [_secureButton addTarget:self action:@selector(secureBtnClicked) forControlEvents:UIControlEventTouchUpInside];
        [_secureButton setImage:[UIImage imageNamed:@"Resource.bundle/eye_close"] forState:UIControlStateNormal];
        [_secureButton setImage:[UIImage imageNamed:@"Resource.bundle/eye_open"] forState:UIControlStateSelected];
        //        _secureButton.selected = YES;
    }
    return _secureButton;
}

@end

File: HungryTools/Classes/UITool/ZHCTextField.swift
//
//  ZHCTextField.swift
//  HungryTools
//
//  Created by 张海川 on 2019/12/25.
//

import UIKit

public class ZHCTextField: UITextField {
    
    public enum TextFieldType {
        case normal
        case number
        case phoneNumber
        case phoneNumberWithoutSpacing
        /// 半角字符 包括字母，数字，标点符号
        case password
        case money
        case idCardNumber
        /// 人名 可以输入汉字、英文字母和数字
        case chinese
        case bankCardNumber
    }
    
    public var fieldType = TextFieldType.normal {
        didSet {
            switch fieldType {
            case .normal:
                break
            case .number:
                keyboardType = .numberPad
            case .phoneNumber:
                keyboardType = .phonePad
                maxLength = 13
            case .phoneNumberWithoutSpacing:
                keyboardType = .phonePad
                maxLength = 11
            case .password:
                isSecureTextEntry = true
                rightView = secureView
                rightViewMode = .always
                keyboardType = .alphabet
                maxLength = 18
            case .money:
                keyboardType = .decimalPad
                maxLength = 8
            case .idCardNumber:
                maxLength = 18
            case .chinese:
                break
            case .bankCardNumber:
                keyboardType = .numberPad
                maxLength = 19
            }
        }
    }
    
    /// 文本长度限制
    public var maxLength = 0
    
    /// 删除空格后的手机号码，在 ANFieldType 为 ANFieldTypePhoneNumber 时有效。
    public var phoneNumberString: String? {
        get {
            if fieldType == .phoneNumber {
                return text?.replacingOccurrences(of: " ", with: "")
            } else {
                return nil
            }
        }
    }
    
    // MARK: - 左视图
    
    /// 左视图文字颜色，需在leftText之前赋值
    public var leftTextColor: UIColor?
    /// 左视图文字字体，需在leftText之前赋值
    public var leftTextFontSize: CGFloat = 14
    /// 左视图文字宽度，需在leftText之前赋值
    public var leftTextWidth: CGFloat?
    /// 左视图文字对齐方式，需在leftText之前赋值
    public var leftTextLabelAlignment: NSTextAlignment = .left
    
    
    public var leftText: String? {
        didSet {
            let label = UILabel()
            label.font = UIFont.systemFont(ofSize: leftTextFontSize)
            if let _ = leftTextColor {
                label.textColor = leftTextColor!
            }
            label.textAlignment = leftTextLabelAlignment
//            let leftView = UIView(frame: CGRect(x: 0, y: 0, width: label.textWidth() + 21, height: bounds.size.height))
            let leftView = UIView()
            
            leftView.addSubview(label)
            addLeftViewConstraints(with: leftView, subView: label)
            if let _ = leftTextWidth {
                leftView.addConstraint(NSLayoutConstraint(item: label,
                                                           attribute: .width,
                                                           relatedBy: .equal,
                                                           toItem: nil,
                                                           attribute: .notAnAttribute,
                                                           multiplier: 1,
                                                           constant: leftTextWidth!))
            }
            
            self.leftView = leftView
            leftViewMode = .always
        }
    }
    
    public var leftImageName: String? {
        didSet {
            guard leftImageName != nil else { return }
            guard let image = UIImage(named: leftImageName!) else { return }
            let subView = UIImageView(image: image)
            
            let leftView = UIView(frame: CGRect(x: 0, y: 0, width: image.size.width + 20, height: bounds.size.height))
            
            leftView.addSubview(subView)
            addLeftViewConstraints(with: leftView, subView: subView)
            
            self.leftView = leftView
            leftViewMode = .always
        }
    }
    
    public override var leftView: UIView? {
        didSet {
            updateBottomLineConstraints()
        }
    }
    
    // MARK: - 右视图
    
    public override var rightView: UIView? {
        didSet {
            updateBottomLineConstraints()
        }
    }
    
    
    // MARK: - 占位符
    
    public var placeholderColor: UIColor? {
        didSet {
            setPlaceHolderAttribute(value: placeholderColor!)
        }
    }
    public var placeholderFont: UIFont? {
        didSet {
            setPlaceHolderAttribute(value: placeholderFont!)
        }
    }
    public override var placeholder: String? {
        didSet {
            if placeholderColor != nil {
                setPlaceHolderAttribute(value: placeholderColor!)
            }
            if placeholderFont != nil {
                setPlaceHolderAttribute(value: placeholderFont!)
            }
        }
    }
    
    
    // MARK: -

    /// 密码明暗文切换图片数组，需包含两个UIImage，第一个代表明文，第二个代表暗文。
    public var secureButtonImages: [UIImage]? {
        didSet {
            guard secureButtonImages?.count == 2 else { return }
            secureButton.setImage(secureButtonImages![0], for: .selected)
            secureButton.setImage(secureButtonImages![1], for: .normal)
        }
    }
    /// clearButton 按钮图片
    public var clearButtonImage: UIImage? {
        didSet {
            let button =  value(forKey: "_clearButton") as! UIButton
            button.setImage(clearButtonImage, for: .normal)
        }
    }
    
    
    
    public var showBottomLine = true {
        didSet {
            bottomLineView.isHidden = !showBottomLine
        }
    }
    /// 正在输入时下划线颜色
    public var bottomLineActiveColor = UIColor.purple
    /// 失去焦点时下划线颜色
    public var bottomLinePassiveColor = UIColor(red: 0.85, green: 0.85, blue: 0.85, alpha: 1) {
        didSet {
            bottomLineView.backgroundColor = bottomLinePassiveColor
        }
    }
    /// 下划线高度，默认0.5
    public var bottomLineHeight: CGFloat = 0.5 {
        didSet {
            removeConstraints(bottomLineView.constraints)
            let margin: CGFloat = isContainLeftOrRightView ? 10 : 0
            addConstraints([
                NSLayoutConstraint(item: bottomLineView, attribute: .left, relatedBy: .equal, toItem: self, attribute: .left, multiplier: 1, constant: margin),
                NSLayoutConstraint(item: bottomLineView, attribute: .bottom, relatedBy: .equal, toItem: self, attribute: .bottom, multiplier: 1, constant: 0),
                NSLayoutConstraint(item: bottomLineView, attribute: .right, relatedBy: .equal, toItem: self, attribute: .right, multiplier: 1, constant: margin),
                NSLayoutConstraint(item: bottomLineView, attribute: .height, relatedBy: .equal, toItem: nil, attribute: .height, multiplier: 1, constant: bottomLineHeight),
            ])
        }
    }
    
    
    
    private lazy var bottomLineView: UIView = {
        let view = UIView()
        view.backgroundColor = bottomLinePassiveColor
        
        return view
    }()
    
    private lazy var secureView: UIView = {
        let view = UIView(frame: CGRect(x: 0, y: 0, width: 64, height: 30))
        
        secureButton.frame = CGRect(x: 10, y: 0, width: 44, height: 30)
        view.addSubview(secureButton)
        
        return view
    }()
    
    private lazy var secureButton: UIButton = {
        let button = UIButton(imageName: "icon_mmbkj_24", target: self, action: #selector(secureBtnClicked))
        button.setImage(UIImage(named: "icon_mmkj_24"), for: .selected)
        
        return button
    }()
    
    private var isContainLeftOrRightView = false
    
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        initConfig()
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
        initConfig()
    }
    
    fileprivate func initConfig() {
        clearButtonMode = .whileEditing
        delegate = self
        font = .systemFont(ofSize: 14)
        
        addBottomLine()
    }
    
    // MARK: - Click Event
    
    @objc private func secureBtnClicked() {
        isSecureTextEntry = !isSecureTextEntry
        secureButton.isSelected = !isSecureTextEntry
    }
}

extension ZHCTextField: UITextFieldDelegate {
    
    public func textField(_ textField: UITextField, shouldChangeCharactersIn range: NSRange, replacementString string: String) -> Bool {
        // 删除
        if string == "" {
            return true
        }
        
        let comingTextLength = (textField.text?.count ?? 0) + string.count - range.length
        if maxLength > 0 && comingTextLength > maxLength {
            return false
        }
        
        switch fieldType {
        case .normal:
            return true
        case .number:
            return string.checkWithRegexString(regex: "[0-9]+")
        case .phoneNumber:
            if textField.text?.count == 3 || textField.text?.count == 8 {
                textField.text?.insert(" ", at: textField.text!.endIndex)
            }
            return string.checkWithRegexString(regex: "[0-9]+")
        case .phoneNumberWithoutSpacing:
            return string.checkWithRegexString(regex: "[0-9]+")
        case .password:
            return string.checkWithRegexString(regex: "[\\x00-\\xff]+") // 半角字符 包括字母，数字，标点符号
        case .money:
            return string.checkWithRegexString(regex: "[0-9.]+")
        case .idCardNumber:
            return string.checkWithRegexString(regex: "[0-9Xx]+")
        case .chinese:
            return string.checkWithRegexString(regex: "[a-z\\u4e00-\\u9fa5]+")
        case .bankCardNumber:
            return string.checkWithRegexString(regex: "[0-9]+")
        }
    }
    
    public func textFieldDidBeginEditing(_ textField: UITextField) {
        UIView.animate(withDuration: 0.25) {
            self.bottomLineView.backgroundColor = self.bottomLineActiveColor
        }
    }
    
    public func textFieldDidEndEditing(_ textField: UITextField) {
        UIView.animate(withDuration: 0.25) {
            self.bottomLineView.backgroundColor = self.bottomLinePassiveColor
        }
    }
}

// MARK: - tools

private extension ZHCTextField {
    
    func addLeftViewConstraints(with leftView: UIView, subView: UIView) {
        subView.translatesAutoresizingMaskIntoConstraints = false
        leftView.addConstraints([
            NSLayoutConstraint(item: subView, attribute: .centerY, relatedBy: .equal, toItem: leftView, attribute: .centerY, multiplier: 1, constant: 0),
            NSLayoutConstraint(item: subView, attribute: .left, relatedBy: .equal, toItem: leftView, attribute: .left, multiplier: 1, constant: 10),
            NSLayoutConstraint(item: subView, attribute: .right, relatedBy: .equal, toItem: leftView, attribute: .right, multiplier: 1, constant: -10),
        ])
    }
    
    func addBottomLine() {
        addSubview(bottomLineView)
        bottomLineView.translatesAutoresizingMaskIntoConstraints = false
        addConstraints([
            NSLayoutConstraint(item: bottomLineView, attribute: .left, relatedBy: .equal, toItem: self, attribute: .left, multiplier: 1, constant: 0),
            NSLayoutConstraint(item: bottomLineView, attribute: .bottom, relatedBy: .equal, toItem: self, attribute: .bottom, multiplier: 1, constant: 0),
            NSLayoutConstraint(item: bottomLineView, attribute: .right, relatedBy: .equal, toItem: self, attribute: .right, multiplier: 1, constant: 0),
            NSLayoutConstraint(item: bottomLineView, attribute: .height, relatedBy: .equal, toItem: nil, attribute: .height, multiplier: 1, constant: bottomLineHeight)
        ])
    }
    
    func updateBottomLineConstraints() {
        
        isContainLeftOrRightView = true
    
        removeConstraints(bottomLineView.constraints)
        addConstraints([
            NSLayoutConstraint(item: bottomLineView, attribute: .left, relatedBy: .equal, toItem: self, attribute: .left, multiplier: 1, constant: 10),
            NSLayoutConstraint(item: bottomLineView, attribute: .bottom, relatedBy: .equal, toItem: self, attribute: .bottom, multiplier: 1, constant: 0),
            NSLayoutConstraint(item: bottomLineView, attribute: .right, relatedBy: .equal, toItem: self, attribute: .right, multiplier: 1, constant: -10),
            NSLayoutConstraint(item: bottomLineView, attribute: .height, relatedBy: .equal, toItem: nil, attribute: .height, multiplier: 1, constant: bottomLineHeight)
        ])
    }
    
    func setPlaceHolderAttribute(value: Any) {
        
        guard attributedPlaceholder != nil else {
            return
        }
        
        let attriString = NSMutableAttributedString(attributedString: attributedPlaceholder!)
        
        var name = NSAttributedString.Key.font
        if value is UIFont {
            name = .font
        } else if value is UIColor {
            name = .foregroundColor
        }
        attriString.addAttribute(name, value: value, range: NSRange(location: 0, length: attributedPlaceholder!.length))
        
        attributedPlaceholder = attriString
    }
}

//private extension String {
//
//    func checkWithRegexString(regex: String) -> Bool {
//        NSPredicate(format: "SELF MATCHES %@", regex).evaluate(with: self)
//    }
//}
